mdsa/pcoa.py

```diff
@@ -0,0 +1,168 @@
+
+    # Run the given algorithm that decomposes the matrix into eigenvectors
+    # and eigenvalues.
+    eigenvectors, eigenvalues = algorithm.run(F_matrix, num_dimensions_out)
+
+    # Generate axis labels for output
+    axis_labels = ['PC%d' % i for i in range(1, eigenvalues.size + 1)]
+
+    # Some algorithms do not return eigenvalues. Thus, we cannot compute
+    # the array of proportion of variance explained either.
+    if np.all(np.isnan(eigenvalues)):
+        # Only return an OrdinationResults object wrapping the result's
+        # eigenvectors. Leave the eigenvalues as NaNs.
+
+        # TODO: do we have to order the eigenvectors..?
+        return OrdinationResults(
+            short_method_name='PCoA',
+            long_method_name='Principal Coordinate Analysis',
+            samples=pd.DataFrame(eigenvectors, index=distance_matrix.ids,
+                                 columns=axis_labels),
+            eigvals=pd.Series(eigenvalues))
+    else:
+        # cogent makes eigenvalues positive by taking the
+        # abs value, but that doesn't seem to be an approach accepted
+        # by Legendre & Legendre to deal with negative eigenvalues.
+        # We raise a warning in that case.
+
+        # First, we coerce values close to 0 to equal 0.
+        indices_close_to_zero = np.isclose(eigenvalues,
+                                           np.zeros(eigenvalues.shape))
+        eigenvalues[indices_close_to_zero] = 0
+
+        if np.any(eigenvalues < 0):
+            warn(
+                "The result contains negative eigenvalues."
+                " Please compare their magnitude with the magnitude of some"
+                " of the largest positive eigenvalues. If the negative ones"
+                " are smaller, it's probably safe to ignore them, but if they"
+                " are large in magnitude, the results won't be useful. See the"
+                " Notes section for more details. The smallest eigenvalue is"
+                " {0} and the largest is {1}.".format(eigenvalues.min(),
+                                                      eigenvalues.max()),
+                RuntimeWarning
+            )
+
+        # eigvals might not be ordered, so we order them (at least one
+        # is zero).
+        indices_descending = eigenvalues.argsort()[::-1]
+        eigenvalues = eigenvalues[indices_descending]
+        # Sort eigenvectors in correspondance with eigenvalues' order
+        eigenvectors = eigenvectors[:, indices_descending]
+
+        # Scale eigenvalues to have lenght = sqrt(eigenvalue). This
+        # works because np.linalg.eigh returns normalized
+        # eigenvectors. Each row contains the coordinates of the
+        # objects in the space of principal coordinates. Note that at
+        # least one eigenvalue is zero because only n-1 axes are
+        # needed to represent n points in an euclidean space.
+        # TODO is this necessary for other algorithms that are NOT eigh?
```

Comment:
Thank you for letting me know. Based on the documentation this is only needed for those methods that return normalized eigenvectors. Thus, we will need to check the documentation of those methods to make sure if they normalize or not those values.